src/Expander.hs

```diff
@@ -1473,8 +1472,37 @@ expandOneForm prob stx
           Id _ -> error "Impossible happened - identifiers are identifier-headed!"
 
 
-expandOneDeclaration :: Scope -> DeclPtr -> Syntax -> DeclValidityPtr -> Expand ()
-expandOneDeclaration sc dest stx ph = expandOneForm (DeclDest dest sc ph) stx
+-- | Each declaration form fills a node in the declaration tree, and also fills
+-- a DeclValidityPtr with the scopes they introduce, so that later code can see
+-- the newly-bound names.
+expandDeclForm :: DeclTreePtr -> DeclValidityPtr -> Syntax -> Expand ()
+expandDeclForm dest vp stx =
+  expandOneForm (DeclTreeDest dest vp) stx
+
+-- | Like a single declaration form, a declaration group both fills a node in
+-- the declaration tree and fills a DeclValidityPtr with the scopes it
+-- introduces. However, since expandDeclForms is recursive, some of those
+-- scopes (those in the given ScopeSet) have already been introduced before
+-- this call. We must merge those with the scopes which will be introduced by
+-- the declarations in the rest of the Syntax.
+expandDeclForms :: DeclTreePtr -> ScopeSet -> DeclValidityPtr -> Syntax -> Expand ()
+expandDeclForms dest scs vp (Syntax (Stx _ _ (List []))) = do
+  linkDeclTree dest DeclTreeLeaf
+  linkDeclValidity vp scs
+expandDeclForms dest scs vp (Syntax (Stx stxScs loc (List (d:ds)))) = do
+  headDest <- liftIO $ newDeclTreePtr
+  tailDest <- liftIO $ newDeclTreePtr
+  headValidityPtr <- newDeclValidityPtr
+  linkDeclTree dest (DeclTreeBranch headDest tailDest)
+
+  -- The head declaration will fill the headValidityPtr with the scopes it
+  -- introduces, and then the ExpandDeclForms task will fill vp with the
+  -- combination of scs, headValidityPtr's scopes, and the scopes indroduced by
+  -- the tail's declarations.
+  expandDeclForm headDest headValidityPtr =<< addRootScope d
+  forkExpanderTask $ ExpandDeclForms tailDest scs headValidityPtr vp (Syntax (Stx stxScs loc (List ds)))
+expandDeclForms _dest _scs _vp _stx =
+  error "TODO real error message - malformed module body"
```

Comment:
In the previous two functions are more comments explaining how I'm accumulating the scopes now.

Reply:
This is really good stuff.